CORE-6685 Standardise error codes for missing virtual nodes

lankydan opened the pull request:
Create `VirtualNodeInfoRPCExtensions` to wrap some existing functions so that HTTP error codes and their messages can be standardised.

lankydan committed:
CORE-6685 Standardise error codes for missing virtual nodes

Create `VirtualNodeInfoRPCExtensions` to wrap some existing functions so
that HTTP error codes and their messages can be standardised.

lankydan commented:
I'm not convinced that `:components:virtual-node:virtual-node-rpcops-common` is the correct place for `VirtualNodeInfoRPCExtensions`, so if anyone has any suggestions, I'm all ears.

vkolomeyko commented:
@lankydan - CI reports build failure:
```

[2022-09-12T08:56:05.136Z] Caused by: org.apache.felix.resolver.reason.ReasonException: Unable to resolve net.corda.flow-rpcops-service-impl version=5.0.0.0-alpha-1662972710329: missing requirement osgi.wiring.package;filter:='(osgi.wiring.package=net.corda.virtualnode.rpcops.common.impl)';osgi.wiring.package='net.corda.virtualnode.rpcops.common.impl'

[2022-09-12T08:56:05.136Z] 	at org.apache.felix.resolver.Candidates$MissingRequirementError.toException(Candidates.java:1341)

[2022-09-12T08:56:05.136Z] 	... 139 more
```

vkolomeyko commented:
> I'm not convinced that `:components:virtual-node:virtual-node-rpcops-common` is the correct place for `VirtualNodeInfoRPCExtensions`, so if anyone has any suggestions, I'm all ears.

This looks sensible to me.

kyriathar commented on components/virtual-node/virtual-node-rpcops-common/src/main/kotlin/net/corda/virtualnode/rpcops/common/impl/VirtualNodeInfoRPCExtensions.kt:
```diff
@@ -0,0 +1,63 @@
+package net.corda.virtualnode.rpcops.common.impl
```
You have also flagged it: This module could become a bucket of conceptually different things. Maybe we should find a better home for these functions, though don't know where myself.

sprintengine replied:
I think it's the correct place, but I think the virtual node sender stuff should be moved to a new module (`virtual-node-management-sender`) and I noted that here:

https://github.com/corda/corda-runtime-os/pull/1943#discussion_r954989314

That would mean things like membership-http-rpc-impl won't have access to virtual node management sender APIs.

So for this PR I think where you put it might be OK.

lankydan replied:
K, I'll rename the current module and extract this code out to a separate module.

lankydan replied:
Done, renamed to `virtual-node-management-sender` and created `virtual-node-info-service-rpc-extensions` for my new code.

kyriathar commented on components/virtual-node/virtual-node-rpcops-common/src/main/kotlin/net/corda/virtualnode/rpcops/common/impl/VirtualNodeInfoRPCExtensions.kt:
```diff
@@ -0,0 +1,63 @@
+ *
+ * Returns `null` if no such information exists
+ */
+fun VirtualNodeInfoReadService.getByHoldingIdentityShortHashOrThrow(holdingIdentityShortHash: ShortHash): VirtualNodeInfo {
+    return getByHoldingIdentityShortHash(holdingIdentityShortHash)
+        ?: throw ResourceNotFoundException("Virtual Node", holdingIdentityShortHash.value)
+}
+
+fun VirtualNodeInfoReadService.getByHoldingIdentityShortHashOrThrow(
+    holdingIdentityShortHash: ShortHash,
+    message: () -> String
+): VirtualNodeInfo {
+    return getByHoldingIdentityShortHash(holdingIdentityShortHash) ?: throw ResourceNotFoundException(message())
+}
+
+/**
+ * Returns the virtual node info by short-hash code, for a given holding identity
+ * without starting any bundles or instantiating any classes.
+ *
+ * Implementation may return null if it is part of a component.
+ *
+ * Short id can be generated by calling the [HoldingIdentity.shortHash]
+ *
+ * Returns `null` if no such information exists
+ */
+fun VirtualNodeInfoReadService.getByHoldingIdentityShortHashOrThrow(holdingIdentityShortHash: String): VirtualNodeInfo {
+    val shortHash = ShortHash.ofOrThrow(holdingIdentityShortHash)
+    return getByHoldingIdentityShortHash(shortHash) ?: throw ResourceNotFoundException("Virtual Node", shortHash.value)
+}
+
+fun VirtualNodeInfoReadService.getByHoldingIdentityShortHashOrThrow(
+    holdingIdentityShortHash: String,
+    message: () -> String
+): VirtualNodeInfo {
+    val shortHash = ShortHash.ofOrThrow(holdingIdentityShortHash)
+    return getByHoldingIdentityShortHash(shortHash) ?: throw ResourceNotFoundException(message())
+}
```
Maybe worth documenting what exception can be thrown and when.

lankydan replied:
Yeah, will do.

kyriathar commented on components/virtual-node/virtual-node-rpcops-common/src/main/kotlin/net/corda/virtualnode/rpcops/common/impl/VirtualNodeInfoRPCExtensions.kt:
```diff
@@ -0,0 +1,63 @@
+ *
+ * Short id can be generated by calling the [HoldingIdentity.shortHash]
+ *
+ * Returns `null` if no such information exists
```
Does this need update? It does not seem it can return null.

kyriathar commented on components/virtual-node/virtual-node-rpcops-common/src/main/kotlin/net/corda/virtualnode/rpcops/common/impl/VirtualNodeInfoRPCExtensions.kt:
```diff
@@ -0,0 +1,63 @@
+ *
+ * Short id can be generated by calling the [HoldingIdentity.shortHash]
+ *
+ * Returns `null` if no such information exists
```
Same as above.

kyriathar commented on components/virtual-node/virtual-node-rpcops-common/src/main/kotlin/net/corda/virtualnode/rpcops/common/impl/VirtualNodeInfoRPCExtensions.kt:
```diff
@@ -0,0 +1,63 @@
+ */
+fun VirtualNodeInfoReadService.getByHoldingIdentityShortHashOrThrow(holdingIdentityShortHash: String): VirtualNodeInfo {
+    val shortHash = ShortHash.ofOrThrow(holdingIdentityShortHash)
+    return getByHoldingIdentityShortHash(shortHash) ?: throw ResourceNotFoundException("Virtual Node", shortHash.value)
```
Perhaps we could change this to call the above overload instead?

lankydan replied:
They use different exception constructors so i cant call the overload.

kyriathar commented on components/virtual-node/virtual-node-rpcops-common/src/main/kotlin/net/corda/virtualnode/rpcops/common/impl/VirtualNodeInfoRPCExtensions.kt:
```diff
@@ -0,0 +1,63 @@
+    message: () -> String
+): VirtualNodeInfo {
+    val shortHash = ShortHash.ofOrThrow(holdingIdentityShortHash)
+    return getByHoldingIdentityShortHash(shortHash) ?: throw ResourceNotFoundException(message())
```
Same as above.

lankydan committed:
CORE-6685 Standardise error codes for missing virtual nodes

Create `VirtualNodeInfoRPCExtensions` to wrap some existing functions so
that HTTP error codes and their messages can be standardised.

Created `virtual-node-info-read-service-rpc-extensions` to house this
code.

Rename `virtual-node-rpcops-common` to `virtual-node-management-sender`.

sprintengine commented on components/virtual-node/virtual-node-management-sender/build.gradle:
```diff
@@ -17,6 +17,9 @@ dependencies {
     implementation project(':libs:lifecycle:lifecycle')
     implementation project(':libs:messaging:messaging')
     implementation project(":libs:utilities")
+    implementation project(':libs:http-rpc:http-rpc')
```
think we can remove these dependencies?

lankydan committed:
CORE-6685 Remove depds

conal-smith reviewed (approved):
lgtm

driessamyn reviewed (approved):
LGTM

ronanbrowne reviewed (approved):
LGTM
